### PR TITLE
[10.x] Cast to string $contents on Component.php to prevent render deprecation messages

### PR DESCRIPTION
**Pull request by eusonlito**
Returning `null` from the render method is a valid option when extending a component:

```php
public function render(): ?View
{
    if ($condition === false) {
        return null;
    }
    
    return view('components.map');
}
```

The above code is still valid; however, since this PR https://github.com/laravel/framework/pull/31271, there's a new method to check if a component should be rendered, called `shouldRender`. Nevertheless, there is code that uses the old way (and it still works).

The issue arises with the `Component.php` file, which is not designed to handle anything other than a `ViewContract` or a `string`.

```php
    /**
     * Create a Blade view with the raw component string content.
     *
     * @param  string  $contents
     * @return string
     */
    protected function extractBladeViewFromString($contents)
    {
```

If `null` is received, the deprecation log reports:

```
[2023-11-07T15:25:53+00:00] deprecations-daily.WARNING: strlen(): Passing null to parameter #1 ($string) of type string is deprecated in /var/www/app/vendor/laravel/framework/src/Illuminate/View/Component.php on line 173
[2023-11-07T15:25:53+00:00] deprecations-daily.WARNING: trim(): Passing null to parameter #1 ($string) of type string is deprecated in /var/www/app/vendor/laravel/framework/src/Illuminate/View/FileViewFinder.php on line 75
[2023-11-07T15:25:53+00:00] deprecations-daily.WARNING: hash(): Passing null to parameter #2 ($data) of type string is deprecated in /var/www/app/vendor/laravel/framework/src/Illuminate/View/Component.php on line 194
```

The solution involves casting the `$contents` variable in Component to a string to prevent these warnings.

### CODE REVIEW

File: src/Illuminate/View/Component.php [hunk 1]
```diff
@@ -147,7 +147,7 @@ public function resolveView()
                 return $view;
             }
 
-            return $this->extractBladeViewFromString($view);
+            return $this->extractBladeViewFromString(strval($view));
```
Comment (driesvints) [inline src/Illuminate/View/Component.php:150]: It's probably best to not invoke the method at all.

```suggestion
            if (is_null($view)) {
                return '';
            }

            return $this->extractBladeViewFromString($view);
```

**Review by driesvints — CHANGES_REQUESTED** [review]

File: src/Illuminate/View/Component.php (hunk 1 above)
Comment (eusonlito) [reply src/Illuminate/View/Component.php:150]: Maybe, as `extractBladeViewFromString` only accepts strings:

```
if (! is_string($view)) {
    return '';
}
```

Comment (eusonlito) [reply src/Illuminate/View/Component.php]: Thanks!

**Comment by driesvints** [comment]
Actually, can't we move this way up in the method? Right after we have the `$view` returned from the `render` method?

**Comment by eusonlito** [comment]
Something like?:

```php
    /**
     * Resolve the Blade view or view file that should be used when rendering the component.
     *
     * @return \Illuminate\Contracts\View\View|\Illuminate\Contracts\Support\Htmlable|\Closure|string
     */
    public function resolveView()
    {
        $view = $this->render();

        if (is_string($view)) {
            return $this->extractBladeViewFromString($view);
        }

        if ($view instanceof ViewContract) {
            return $view;
        }

        if ($view instanceof Htmlable) {
            return $view;
        }

        $resolver = function ($view) {
            if ($view instanceof ViewContract) {
                return $view;
            }

            return $this->extractBladeViewFromString($view);
        };

        if ($view instanceof Closure) {
            return fn (array $data = []) => $resolver($view($data));
        }

        return '';
    }
```

**Comment by driesvints** [comment]
No like this:

```
        $view = $this->render();

        if (! is_null($view)) {
            return '';
        }
```

**Comment by eusonlito** [comment]
But we are again on `is_null` instead `is_string`.

I think that we can simplify the function as:

```php
    /**
     * Resolve the Blade view or view file that should be used when rendering the component.
     *
     * @return \Illuminate\Contracts\View\View|\Illuminate\Contracts\Support\Htmlable|\Closure|string
     */
    public function resolveView()
    {
        $resolver = function ($view) {
            if (($view instanceof ViewContract) || ($view instanceof Htmlable)) {
                return $view;
            }

            if (is_string($view)) {
                return $this->extractBladeViewFromString($view);
            }

            return '';
        };

        $view = $this->render();

        if ($view instanceof Closure) {
            return fn (array $data = []) => $resolver($view($data));
        }

        return $resolver($view);
    }
```

Original method is (as reference):

```php
    /**
     * Resolve the Blade view or view file that should be used when rendering the component.
     *
     * @return \Illuminate\Contracts\View\View|\Illuminate\Contracts\Support\Htmlable|\Closure|string
     */
    public function resolveView()
    {
        $view = $this->render();

        if ($view instanceof ViewContract) {
            return $view;
        }

        if ($view instanceof Htmlable) {
            return $view;
        }

        $resolver = function ($view) {
            if ($view instanceof ViewContract) {
                return $view;
            }

            return $this->extractBladeViewFromString($view);
        };

        return $view instanceof Closure ? function (array $data = []) use ($view, $resolver) {
            return $resolver($view($data));
        }
        : $resolver($view);
    }
```

**Review by driesvints — APPROVED** [review]

**Comment by driesvints** [comment]
Lgtm!

**Comment by taylorotwell** [comment]
I dunno, I worry this could be a breaking change to people returning other Stringable objects, etc. Just return a string.

**Comment by driesvints** [comment]
Actually @eusonlito. Upon further digging I don't think `render` can ever return `null`. There's no example of that in https://laravel.com/docs/10.x/blade and the return annotation is clear about this. What makes you think it can?

https://github.com/laravel/framework/blob/10.x/src/Illuminate/View/Component.php#L81

**Comment by eusonlito** [comment]
As noted in the PR description, prior to this PR https://github.com/laravel/framework/pull/31271, the method to prevent a component from being rendered involved passing null or an empty string as the render argument. Since there were no deprecation warnings at these time, this approach simply worked. However, you are correct; it wasn't the proper way to handle it.

Thank you for the review.

**Pull request closed without merging**